Sources/ServiceDiscovery/ServiceDiscovery.swift

```diff
@@ -12,120 +12,42 @@
-    ///
     /// - Parameters:
     ///   - service: The service to lookup
-    ///   - deadline: Lookup is considered to have timed out if it does not complete by this time
-    ///   - callback: The closure to receive lookup result
-    func lookup(_ service: Service, deadline: DispatchTime?, callback: @escaping (Result<[Instance], Error>) -> Void)
+    ///
+    /// -  Returns: A listing of service instances.
+    func lookup(_ service: Service, deadline: ContinuousClock.Instant?) async throws -> [Instance]
 
     /// Subscribes to receive a service's instances whenever they change.
     ///
-    /// The service's current list of instances will be sent to `nextResultHandler` when this method is first called. Subsequently,
-    /// `nextResultHandler` will only be invoked when the `service`'s instances change.
-    ///
-    /// ### Threading
-    ///
-    /// `nextResultHandler` and `completionHandler` may be invoked on arbitrary threads, as determined by implementation.
+    /// Returns a ``ServiceDiscoveryInstanceSequence``, which is an `AsyncSequence` and each of its items is a snapshot listing of service instances.
     ///
     /// - Parameters:
     ///   - service: The service to subscribe to
-    ///   - nextResultHandler: The closure to receive update result
-    ///   - completionHandler: The closure to invoke when the subscription completes (e.g., when the `ServiceDiscovery` instance exits, etc.),
-    ///                 including cancellation requested through `CancellationToken`.
     ///
-    /// -  Returns: A ``CancellationToken`` instance that can be used to cancel the subscription in the future.
-    func subscribe(to service: Service, onNext nextResultHandler: @escaping (Result<[Instance], Error>) -> Void, onComplete completionHandler: @escaping (CompletionReason) -> Void) -> CancellationToken
-}
-
-// MARK: - Subscription
-
-/// Enables cancellation of service discovery subscription.
-public class CancellationToken {
-    private let _isCancelled: ManagedAtomic<Bool>
-    private let _completionHandler: (CompletionReason) -> Void
-
-    /// Returns `true` if the subscription has been cancelled.
-    public var isCancelled: Bool {
-        self._isCancelled.load(ordering: .acquiring)
-    }
-
-    /// Creates a new token.
-    public init(isCancelled: Bool = false, completionHandler: @escaping (CompletionReason) -> Void = { _ in }) {
-        self._isCancelled = ManagedAtomic<Bool>(isCancelled)
-        self._completionHandler = completionHandler
-    }
-
-    /// Cancels the subscription.
-    public func cancel() {
-        guard self._isCancelled.compareExchange(expected: false, desired: true, ordering: .acquiring).exchanged else { return }
-        self._completionHandler(.cancellationRequested)
-    }
+    /// -  Returns: A ``ServiceDiscoveryInstanceSequence`` async sequence.
+    func subscribe(_ service: Service) async throws -> any ServiceDiscoveryInstanceSequence<Instance>
```

Comment:
Now the sequence items are a service's individual items rather than snapshots of instance list. How would the consumer know when an instance is removed?

Reply:
this is an interesting question. I suppose we implementation would manage the state and provide atomic (added, deleted) updates. @ktoso do you have a preference one way to the other? ie do you think the client of this would prefer atomic changes or getting the set of instance that are "current"?

Reply:
addressed